### DISCUSSION TITLE
Very new user: Calculate the cartesian atomic coordinates for a molecule

### DISCUSSION THREAD

**Discussion by a-jp** [openmopac/mopac#189] [Q&A]
All,

I would like to calculate for a given molecule the cartesian atomic coordinates in Angstroms.  Would it be possible to have a working example to do this that would allow me to vary the input molecule and get as an output, ideally in a simple text file, the above?

I would very much appreciate any help here, I am new to the code.  I appreciate I will need to understand eventually aspects relating to optimisation options and the model used PM3/PM7.  But a working example with trusted defaults would be most welcome.

Many thanks,

**Comment by nbehrnd** [comment]
What is the file format your molecule of interest is described?  Because there are a couple of them, the most elementary the .xyz, where atoms' positions are described in a Cartesian coordinate system with (an implicit) scale in Angstrom.  See e.g., [openbabel](https://open-babel.readthedocs.io/en/latest/FileFormats/XYZ_cartesian_coordinates_format.html).

If "calculate for a given molecule the cartesian atomic coordinates" refers to a conversion of the underlying file format, [openbabel](https://github.com/openbabel/openbabel) (with its optional [gui](https://open-babel.readthedocs.io/en/latest/GUI/GUI.html)) might suffice.  Examples of conversion are compiled [here](https://open-babel.readthedocs.io/en/latest/Command-line_tools/babel.html#examples).

If you seek a GUI to set up a subsequent computation for instance with MOPAC, [avogadro](https://two.avogadro.cc/) and its [input generators](https://two.avogadro.cc/docs/whats-new-in-avogadro-2/interface-changes.html#input-generators) is one example of a helpful assistant:

![example_for_mopac](https://github.com/openmopac/mopac/assets/19709084/46c22b76-b518-49de-915f-ec0f29fcf5a7)

**Reply by a-jp** [reply]
Would anyone have a sample input file for me to play with that computes the cartesian atomic coordinates? Thanks

**Reply by godotalgorithm** [reply]
My reply below contains a simple example, but you need both OpenBabel and MOPAC installed. Also, the main input file is for OpenBabel, which then generates the input file for MOPAC.

**Comment by godotalgorithm** [comment]
This is more of a discussion than an issue, so I've moved it over accordingly. Also, I received a reply from @a-jp by email that didn't appear in the thread, so either they deleted it or replied in a way that didn't show up online for some reason. In that reply, they asked about what to do when you want to study a molecule with a tool like MOPAC but don't have it represented in any standard input file format.

To be clear, while MOPAC has some flexibility in input file formats ([manual entry](http://openmopac.net/manual/geometry_specification.html)), you still need to provide an approximate geometry of a molecule as input for MOPAC to produce any useful output. With so much chemistry information online and powerful conversion tools like [OpenBabel](http://openbabel.org), there are many ways to generate a useable MOPAC input file.

@nbehrnd's reply describes a very popular approach to generating an approximate geometry: you use molecular drawing software like [Avogadro](https://avogadro.cc) to draw a rough 3D sketch of the molecule you are interested in and then save it in a format directly compatible with MOPAC. However, that requires you to understand a particular molecule enough to draw it and be familiar enough with a drawing program to know how to draw it, so there is a learning curve and a knowledge check.

There is a simpler alternative for well understood molecules that allows you to avoid having to draw it yourself. Online sources of information about molecules like Wikipedia or PubChem usually have some downloadable descriptions of a molecule that can be readily converted to a MOPAC input. Here is a 3-step process to get you started:

1. Find the PubChem entry for the molecule that you are interested in, such as methane ([here](https://pubchem.ncbi.nlm.nih.gov/compound/297)), and save one of the OpenBabel-compatible molecular descriptions to a file. OpenBabel can work with a 2D structure, a 3D structure, an InChI string (saved as a `*.inchi` text file), or a SMILES string (saved as a `*.smiles` text file). The simplest option for this example is to create a text file named `methane.smiles` which contains a single line of text: `C`.

2. Convert to a MOPAC input file using OpenBabel. If you have the command-line version of OpenBabel installed, you would run the command `obabel -ismiles methane.smiles -omop --gen3d -O methane.mop` in the directory containing the `methane.smiles` file. This syntax is explained in the [OpenBabel documentation](https://open-babel.readthedocs.io). In particular, OpenBabel has a system for guessing 3D structures from descriptions like SMILES that only describe how atoms are connected.

3. Add keywords to the MOPAC input file and run it. The first line of `methane.mop` generated by OpenBabel contains `PUT KEYWORDS HERE`. You need to replace that text with some valid MOPAC keywords, such as `PM3` if you want to use that older model or `PM7`, which is also the default if you don't specify a model. With some proper keywords added or even a blank keyword line, you then run MOPAC for this example on the command line as `mopac methane.mop` and the output will be saved in the file `methane.out`.

I hope this helps, and feel free to ask more clarifying questions.

**Reply by a-jp** [reply]
Thanks for your help. How would you advise that I proceed to obtain the atomic coordinates from mopac if I do not have, or cannot find, the initial information I need? Where do I start? And with which program to eventually end back at mopac?

**Reply by godotalgorithm** [reply]
Can you give me some examples of what you know about a molecule that you would like to convert to atomic coordinates?

If you have a standard chemical name, then that can usually lead you to a unique PubChem entry ([starting from here](https://pubchem.ncbi.nlm.nih.gov)), and then you can follow the instructions from this example. You can also search PubChem with just a chemical formula, but multiple isomers might have the same chemical formula. You'll need to know more information than just a chemical formula to narrow it down from there.

Unless you are using a tool or database that can provide MOPAC-compatible files directly, you will almost certain need to install and use OpenBabel to convert some other file type that describes your molecule to a MOPAC input file.

**Reply by nbehrnd** [reply]
For a selection of (somewhat simpler) molecules, an other approach to access them is the [Wikipedia Chemical Structure Explorer](https://wikipedia.cheminfo.org/).  Here, one can query the English edition of the encyclopedia by chemical/commercial name or (sub)structure/similar structure to be directed to the corresponding pages in Wikipedia itself which usually feature a property box.  From here, one can fetch e.g., the SMILES string

![example_aspirine](https://github.com/openmopac/mopac/assets/19709084/a1821f8a-e9cf-46db-98f3-d75a97382b17)

which can be a starting point for openbabel (here `O=C(C)Oc1ccccc1C(=O)O`).  With a working installation of openbabel, you then can generate e.g., a .xyz file by the command to the command line

```shell
obabel -:"O=C(C)Oc1ccccc1C(=O)O" --gen3d -h -O structure.xyz
```

which offers one reasonable conformer

```shell
$ cat structure.xyz 
21

O          2.68963       -1.34199        0.03422
C          2.21264       -0.26327       -0.29484
C          0.73887        0.01692       -0.30573
O          2.93209        0.88139       -0.67661
C          4.35135        0.89692       -0.80712
C          5.06348       -0.25422       -1.16702
C          6.45421       -0.21249       -1.30370
C          7.14252        0.98310       -1.11964
C          6.43643        2.14092       -0.79066
C          5.03813        2.11255       -0.63212
C          4.39931        3.40346       -0.25763
O          4.99699        4.27846        0.34811
O          3.13859        3.56358       -0.70343
H          0.19516       -0.93644       -0.31039
H          0.45126        0.55971       -1.20953
H          0.45490        0.57740        0.58905
H          4.56703       -1.20416       -1.34236
H          7.00301       -1.11851       -1.55391
H          8.22235        1.01539       -1.22592
H          6.98406        3.07302       -0.66057
H          2.80735        2.81189       -1.22837
```

**Reply by godotalgorithm** [reply]
For such Wikipedia entries of molecules with a "Chemical and physical data" block, you can also get direct access to atomic coordinates by clicking on the JSMol interactive image link. From the JSMol viewer, right-click the mouse and go to "Show" and then "Extract MOL data". The MOL file that appears in the text box isn't very common, but it can also be converted by OpenBabel to a more useful format like a MOPAC input file. This avoids having to rely on OpenBabel to guess the atomic coordinates from connectivity information.

**Reply by a-jp** [reply]
Thank you both for your help, very much appreciated!  I must ask out of curiosity, if we cannot find entries of the molecule in any of the cited locations above - how would you proceed?  What codebase would you use to get started?

**Reply by nbehrnd** [reply]
@godotalgorithm Yes, the presence of interactive JSMol of [jmol](https://jmol.sourceforge.net/) including its console is a nice (more recent) addition to Wikipedia.

**Reply by nbehrnd** [reply]
@a-jp There are many large databases, for instance the ones by [SpringerMaterials](https://www.springernature.com/de/librarians/products/databases-solutions/springermaterials); crystal data repositories like the [CSD](https://www.ccdc.cam.ac.uk/solutions/software/csd/), [ISCD](https://icsd.products.fiz-karlsruhe.de/), [COD](http://www.crystallography.net/cod/). The interfaces by [Reaxys](https://www.reaxys.com/) (by Elsevier), or Scifinder/Chemical Abstracts (by The American Chemical Society) equally allow you to access experimentally determined structures, or a generic .mol file which can be processed further.

With *code basis*, you might be interested in *structure generation* of molecules potentially not yet synthesized.  Programs in this field include [SMOG](https://pubs.acs.org/doi/10.1021/ci950393z), [molgen](http://molgen.de/) and the [Open Molecule Generator (OMG)](https://jcheminf.biomedcentral.com/articles/10.1186/1758-2946-4-21).  Examples of the systematic exploration of "the chemical space" of organic molecules include the [GDB project](https://gdb.unibe.ch/research/) by the Reymond group, or [QM9](http://quantum-machine.org/) by von Lilienfeld *et al.* equally worth a literature search as is to search repositories about `molecule` and `generate` / `generation` here on GitHub (not an endorsement, but an [example](https://github.com/kevinid/molecule_generator)).

**Reply by godotalgorithm** [reply]
As you may have noticed from many of these examples, you often have to combine multiple resources and pieces of software to complete tasks in computational chemistry. If you want a recommendation on learning a "codebase" that tries to combine and connect the utility of many tools and resources in computational chemistry, I would recommend learning about the Python programming language. Tasks that require multiple files being passed between multiple programs through the command line can be automated and generalized with Python scripting. Many programs also have their own Python interfaces ([OpenBabel](https://open-babel.readthedocs.io/en/latest/UseTheLibrary/PythonDoc.html), for example) and some online resources even have Python interfaces ([Cambridge Structure Database](https://www.ccdc.cam.ac.uk/solutions/software/csd-python/), for example), which makes connecting them together even easier. Finally, Python has a good multi-platform package manager, [Anaconda](https://anaconda.org), which can automate the installation of all the software you would need to run a complicated Python script that chains together many programs.

The caveat to using Python is that you do need to learn how to program at a basic level, which is more of an investment than learning to work on the command line, and much more of an investment than working with a program that has a graphical interface. At least there are abundant online resource for learning chemistry: for example, the institute that I work for has some [online resources](https://education.molssi.org/python_scripting_cms/) for learning Python that are designed for chemists.

In my experience, computational chemistry work is usually very polarized. Either you have a simple task and the easiest thing to do is to draw a molecule "by hand" using tools like [Avogadro](https://avogadro.cc) to generate its atomic coordinates, or you have a very complicated task that is probably best organized through Python scripts or some other scripting language so that you don't forget what you did. Drawing a molecule in a 3D molecular viewer is the digital analog of building a ball-and-stick model, and I find such an exercise to be instructive and rewarding in its own right, in addition to providing you with the data to feed into simulation software like MOPAC.

All of the software and resources suggested in this thread are open-source and openly accessible. There is also commercial software, which can potentially be a lot more user-friendly with more emphasis on graphical interfaces in exchange for having to pay for it. I am not familiar enough with that space to make recommendations or say whether they are worth the cost, but the usual suspects there are the software products of [Schrodinger](https://www.schrodinger.com), [BIOVIA](https://www.3ds.com/products/biovia), [OpenEye](https://www.eyesopen.com), and [Materials Design](https://www.materialsdesign.com). My general view on these products is that they expand the set of tasks that you can perform easily, but you will still need to learning programming (or get someone to program for you) to perform more complicated tasks in computational chemistry.
